org.eclipse.jdt.launching/launching/org/eclipse/jdt/launching/JavaRuntime.java

```diff
@@ -3460,6 +3465,104 @@ public static VMStandin createVMFromDefinitionFile(File eeFile, String name, Str
 		throw new CoreException(status);
 	}
 
+	/**
+	 * Returns the names of all packages publicly exported by the given {@link IVMInstall vmInstall}, optionally restricted to the given release
+	 * version.
+	 * <p>
+	 * The returned set contains the distinct names sorted in alphabetical order.<br>
+	 * The specified release is only considered for {@link #isModularJava(IVMInstall) modular} VM-Installs, for non-modular installs it is ignored,
+	 * and then must correspond to Java version 9 or higher.
+	 * </p>
+	 *
+	 * @param vm
+	 *            the vm-install to query
+	 * @param release
+	 *            the release to restrict the packages to, ignored for non-modular VMs. Supported values are the constants defined in
+	 *            {@link JavaCore#VERSION_9 JavaCore.VERSION_X}, e.g. {@code "9"}, {@code "10"}, {@code "11"} or so on. If {@code null}, the
+	 *            VMInstall's release is used.
+	 * @return the distinct and alphabetically sorted immutable set of names of all packages in the VM
+	 * @throws CoreException
+	 *             if an error occurs or a modular VM is queried for packages of Java-8 or lower
+	 * @since 3.22
+	 */
+	@SuppressWarnings("restriction")
+	public static Set<String> getProvidedVMPackages(IVMInstall vm, String release) throws CoreException {
+		boolean isModular = JavaRuntime.isModularJava(vm);
+		if (release != null) {
+			// normalize release (from 17.0.6 to 17)
+			release = org.eclipse.jdt.internal.compiler.impl.CompilerOptions.versionFromJdkLevel(org.eclipse.jdt.internal.compiler.impl.CompilerOptions.versionToJdkLevel(release));
+		}
+		// If release is empty we assume no release info available
+		if (release != null && release.isEmpty()) {
+			release = null;
+		}
```

Comment:
I wonder if we should just throw an exception if the `release` value is not as expected? We could even consider to throw one if a non modular `release`, i.e. `1.8` or less is specified since they are effectively ignored (but that might require too much logic on the call side)?
If I understand the code called correctly it will result in an empty string and eventually `null` if one just specifies any non-sense.

Reply:
May I ask you to try with Java 23 release (I haven't tried)? JDT doesn't fully support it yet, so I believe the code above will return empty string, and the question would be: if so, would it make sense to continue (assuming it is the JVM we are running Eclipse) or throw an error? If we continue, would the client of this API get what is expected or not?

Reply:
If I call for example `JavaRuntime.getProvidedVMPackages(vm, "24")`
Then the following code is called:
https://github.com/eclipse-jdt/eclipse.jdt.core/blob/279fc9d3e2dfbf00a30b2046ee710dfdc84aaeab/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/impl/CompilerOptions.java#L909-L912

Which eventually returns `ClassFileConstants.MAJOR_LATEST_VERSION` that currently corresponds to Java 22.
So that does not fail.
Thinking about it again, I think a suitable compromise would be to throw an exception if `CompilerOptions.versionToJdkLevel()` returns 0. This would permit JVM versions like `17.0.10` which is then mapped to spec verison `17`, but it would reject random other strings as input.

The only problem I see with this approach is that 23 is currently mapped to 22, even if the targeted JVM is e.g. a 23 EA build and could be queried. We should actually just check if the passed release argument is a version and leave the check if that version is supported to the JRT filesystem.

Reply:
> a suitable compromise would be to throw an exception if CompilerOptions.versionToJdkLevel() returns 0.

My concern is that people would not be able to use this API running on latest JVM not yet supported by JDT.
I assume you plan to use it in PDE, so what would break in that case, assuming they "target" JVM will be "known" by JDT, so that in theory we shouldn't fail in this case?

Reply:
> > a suitable compromise would be to throw an exception if CompilerOptions.versionToJdkLevel() returns 0.
> 
> My concern is that people would not be able to use this API running on latest JVM not yet supported by JDT. I assume you plan to use it in PDE, so what would break in that case, assuming they "target" JVM will be "known" by JDT, so that in theory we shouldn't fail in this case?

Yes my plan is to use that in PDE and in general if querying the System-Packages fails it would either mean that for the corresponding EE no system-packages are available or we just not provide the EE at all (the exact behavior is to be defined). At the moment only the EEs defined by JDT are available in PDE's state anyway.
That being said, for PDE this new API should be called in a way that is always correct and never throws an exception (at this point).
My intention for throwing an exception is for yet not known other users that should get some feed back if they pass unsuitable values.

But with the current approach fetching the system-packages for Java-23 is possible even if one has a JDK 23 installed, because 23 is converted to 22 as described before although the `JavaSE-23` EE already exists: https://github.com/eclipse-jdt/eclipse.jdt.debug/pull/426
Therefore the current approach is not suitable in this case.

I'm currently thinking about a more fine tuned logic.

Reply:
I have just pushed a second commit to this branch with a suggestion that should prevent the biggest mistakes but still allows to query JDKs of version that are not yet supported by ECJ/java-core.
The second commit also contains corresponding adjustments/extensions of the tests.

If you are fine with them that commit should be squashed into the first.

Reply:
Yes, I think it is OK, so please squash